content/4_Gold/DSU.mdx

```diff
@@ -86,106 +85,146 @@ $x$ otherwise.
 </Resources>
 
 ```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
 struct DSU {
-	vi e; void init(int N) { e = vi(N,-1); }
-	// get representive component, uses path compression
+	vector<int> e;
+	DSU(int N) { e = vector<int>(N, -1); }
+
+	// get representive component (uses path compression)
 	int get(int x) { return e[x] < 0 ? x : e[x] = get(e[x]); }
-	bool sameSet(int a, int b) { return get(a) == get(b); }
+	
+	bool same_set(int a, int b) { return get(a) == get(b); }
+	
 	int size(int x) { return -e[get(x)]; }
-	bool unite(int x, int y) { // union by size
-		x = get(x), y = get(y); if (x == y) return 0;
-		if (e[x] > e[y]) swap(x,y);
-		e[x] += e[y]; e[y] = x; return 1;
+	
+	bool unite(int x, int y) {  // union by size
+		x = get(x), y = get(y);
+		if (x == y) return false;
+		if (e[x] > e[y]) swap(x, y);
+		e[x] += e[y]; e[y] = x; return true;
 	}
 };
 ```
 
 </CPPSection>
-
 <JavaSection>
 
-This implementation assumes that `disjoint` is an array that starts such that
-`disjoint[i] = -1` for every $0 <= i < n$.
-
 ```java
-public static int disjoint[]; //0 indexed
-public static int size[];
-//finds the "representative" node in a's component
-public static int find(int v) {
-	if (disjoint[v] == -1) {
-		return v;
+import java.util.Arrays;
+
+public class DisjointSets {
+	int[] parents;  // 0-indexed
+	int[] sizes;
+	public DisjointSets(int size) {
+		sizes = new int[size];
+		parents = new int[size];
+		Arrays.fill(sizes, 1);
+		Arrays.fill(parents, -1);
 	}
-	disjoint[v] = find(disjoint[v]);
-	return disjoint[v];
-}
 
-//returns whether the merge changed connectivity
-public static void union(int u, int v) {
-	int uroot = find(u);
-	int vroot = find(v);
-	if (uroot == vroot) {
-		return;
+	// finds the "representative" node in a's component
+	public int find(int x) {
+		return parents[x] == -1 ? x : (parents[x] = find(parents[x]));
 	}
-	if (size[uroot] < size[vroot]) {
-		disjoint[uroot] = vroot;
-		size[vroot] = size[uroot] + size[vroot];
-	} else {
-		disjoint[vroot] = uroot;
-		size[uroot] = size[uroot] + size[vroot];
+
+	// returns whether the merge changed connectivity
+	public boolean union(int x, int y) {
+		int xRoot = find(x);
+		int yRoot = find(y);
+		if (xRoot == yRoot) {
+			return false;
+		}
+		if (sizes[xRoot] < sizes[yRoot]) {
+			parents[xRoot] = yRoot;
+			sizes[yRoot] = sizes[xRoot] + sizes[yRoot];
+		} else {
+			parents[yRoot] = xRoot;
+			sizes[xRoot] = sizes[xRoot] + sizes[yRoot];
+		}
+		return true;
 	}
 }
 ```
 
 </JavaSection>
+<PySection>
+
+```py
+class DisjointSets:
+	def __init__(self, size: int) -> None:
+		self.parents = [-1 for _ in range(size)]
```

Comment:
also up to you
```suggestion
		self.parents = [-1] * size
```

Reply:
not really necessary tbh, and not like it helps readability